interp/constrextern.ml

```diff
@@ -721,15 +721,27 @@ let extern_possible_prim_token (custom,scopes) r =
    | None -> raise No_match
    | Some key -> insert_coercion coercion (insert_delimiters (CAst.make ?loc:(loc_of_glob_constr r) @@ CPrim n) key)
 
-let extern_possible extern r =
-  try Some (extern r) with No_match -> None
-
-let extern_optimal extern r r' =
-  let c = extern_possible extern r in
-  let c' = if r==r' then None else extern_possible extern r' in
-  match c,c' with
-  | Some n, (Some ({ CAst.v = CDelimiters _}) | None) | _, Some n -> n
-  | _ -> raise No_match
+let filter_fully_applied r l =
+  let nargs = match DAst.get r with
+  | GApp (f,args) -> List.length args
+  | _ -> assert false in
+  List.filter (fun (keyrule,pat,n as _rule) -> n = Some nargs) l
+
+let extern_optimal extern extern_fully_applied r r' =
+  if r==r' then
```

Comment:
Use of physical equality is very suspicious here, right?

Reply:
Ping on this comment @herbelin

Reply:
It is an optimisation (which accordingly assumes that extern_coercion returns the original pointer if no coercions are removed). Would you prefer that a `Some` is returned in case of a change?

Reply:
You know physical equality is a bit of a slippery slope in functional languages, moreover then a test / invariant should be added to the documentation of `extern_coercion` so it preserves it; I guess using `Some` would be more future-proof?

Reply:
For example, that cannot happen today [I believe], but it could be the case than it the future the GC would trigger and decided to compact the heap thus making the above equality fail. Thus `==`  seems only safe to detect when two objects _are not_ equal, if we want to check for equality we are bound to call `=` after `==` succeeds.

Reply:
This is not true, since the Gc does not duplicate, it *atomically* moves.

Reply:
Removing the `==` happens to add some complexity and I prefer to renounce. Anyway, if the equality fails, this only impact complexity. At worst, we redo the search for a notation twice.

Moreover, in the next refinement, #11172, `==` is anyway removed.

Reply:
Indeed, we discussed offline and we can have this tradeoff for 8.10 / 8.11 and clean up more in 8.12.